src/sentry/processing/realtime_metrics/redis.py

```diff
@@ -0,0 +1,55 @@
+import datetime
+from typing import Union
+
+from sentry.exceptions import InvalidConfiguration
+from sentry.utils import redis
+
+from . import base
+
+
+class RedisRealtimeMetricsStore(base.RealtimeMetricsStore):
+    """An implementation of RealtimeMetricsStore based on a Redis backend."""
+
+    def __init__(
+        self,
+        cluster: str,
+        counter_bucket_size: int,
+        counter_ttl: datetime.timedelta,
+    ) -> None:
+        """Creates a RedisRealtimeMetricsStore.
+
+        "cluster" is the name of the Redis cluster to use. "counter_bucket_size" is the size
+        in second of the buckets that timestamps will be sorted into when a project's counter is incremented.
+        "counter_ttl" is the duration that counter entries will be kept around for *after
+        the last increment call*.
+        """
+
+        self.cluster = redis.redis_clusters.get(cluster)
+        self._counter_bucket_size = counter_bucket_size
+        self._counter_ttl = int(counter_ttl / datetime.timedelta(milliseconds=1))
+        self._prefix = "symbolicate_event_low_priority"
+
+    def validate(self) -> None:
+        if self._counter_bucket_size <= 0:
+            raise InvalidConfiguration("bucket size must be at least 1")
+
+    def increment_project_event_counter(
+        self, project_id: int, timestamp: Union[int, float]
+    ) -> None:
+        """Increment the event counter for the given project_id.
+
+        The counter is used to track the rate of events for the project.
+        Calling this increments the counter of the current
+        time-window bucket with "timestamp" providing the time of the event.
+        """
+
+        timestamp = int(timestamp)
+        if self._counter_bucket_size > 1:
+            timestamp -= timestamp % self._counter_bucket_size
+
+        key = f"{self._prefix}:counter:{self._counter_bucket_size}:{project_id}:{timestamp}"
+
+        with self.cluster.pipeline() as pipeline:
+            pipeline.incr(key)
+            pipeline.pexpire(key, self._counter_ttl)
```

Comment:
should this be set to `self._counter_ttl + (self._counter_bucket_size - timestamp % self._counter_bucket_size)`? in other words, should it be `time to live + amount of time left until the current bucket rolls over to the next one` to ensure that the key lives TTL _longer_ than the end of the bucket?

if i have a bucket that lives from 0-10 seconds and i create a key at 6 seconds, then the TTL should be 300 seconds _after_ the 10 second mark since we care about the buckets and not the specific time when the key was incremented, right? i think if we can do this then we should be able to just set the expiration on this once if the key doesn't exist and call it a day. this might also solve the hashmap expiration issue as well; thought?

Reply:
The problem is that in the hashmap case there is no way that I can find to set the expiry only once. The only choices are to not set an expiration at all or set it every time. Therefore, if we want the time at which the bucket expires to be fixed a priori, we need to either compute the remaining time until the expiry point and set that with `expire` or compute the the expiry point and set it with `expireat`. The latter option has the disadvantage of being hard to test.

Question is, does it matter if we know exactly when a bucket expires? Otherwise we can just set the expiry to `ttl + bucket_size` on every update; the bucket is then guaranteed to live `ttl` seconds after its upper bound.

Reply:
> Question is, does it matter if we know exactly when a bucket expires?

i believe that it is important that we know exactly when a bucket expires, mostly because it'll affect some work that relies on bucket expiry as a means of filtering out metrics. i don't think it makes sense to block this PR on the specific detail of what the expiration date is though, as setting the expiration to some known value is enough for the scope of this PR. my apologies for that!